ta/pkcs11/src/token_capabilities.c

```diff
@@ -0,0 +1,240 @@
+#endif
+
+#define MECHA(_label, _dig, _enc, _dec, _sig, _ver,		\
+	       _sr, _vr, _der, _wra, _unw, _gen, _gpa, _1s)	\
+	MECHANISM(PKCS11_CKM_ ## _label,				\
+		  (_enc ? PKCS11_CKFM_ENCRYPT : 0) |		\
+		  (_dec ? PKCS11_CKFM_DECRYPT : 0) |		\
+		  (_dig ? PKCS11_CKFM_DIGEST : 0) |		\
+		  (_sig ? PKCS11_CKFM_SIGN : 0) |			\
+		  (_sr ? PKCS11_CKFM_SIGN_RECOVER : 0) |		\
+		  (_ver ? PKCS11_CKFM_VERIFY : 0) |		\
+		  (_vr ? PKCS11_CKFM_VERIFY_RECOVER : 0) |	\
+		  (_gen ? PKCS11_CKFM_GENERATE : 0) |		\
+		  (_gpa ? PKCS11_CKFM_GENERATE_KEY_PAIR : 0) |	\
+		  (_wra ? PKCS11_CKFM_WRAP : 0) |		\
+		  (_unw ? PKCS11_CKFM_UNWRAP : 0) |		\
+		  (_der ? PKCS11_CKFM_DERIVE : 0),		\
+		  _1s)
+
+static const struct pkcs11_mechachism_modes pkcs11_modes[] = {
+	/*
+	 * PKCS#11 directives on mechanism support for the several processing
+	 * modes.
+	 *				1: One shot processing supported ---.
+	 *				Gp: Generate secret pair --------.  |
+	 *				Ge: Generate secret value ----.  |  |
+	 *				Wr|Uw: Wrap/Unwrap -------.   |  |  |
+	 *				Dr: Derive ----------.    |   |  |  |
+	 *		Sr|Vr: SignRecover/VerifyRecov --.   |    |   |  |  |
+	 *		Si|Ve: Sign/Verify --------.     |   |    |   |  |  |
+	 *		En|De: Encrypt/Decrypt     |     |   |    |   |  |  |
+	 *		Di: Digest -----.    |     |     |   |    |   |  |  |
+	 *				|   / \   / \   / \  |   / \  |  |  |
+	 * Mechanism			Di|En|De|Si|Ve|Sr|Vr|Dr|Wr|Uw|Ge|Gp|1
+	 */
+	MECHA(AES_ECB,			0, 1, 1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0),
+	/*
+	 * Mechanism			 Di|En|De|Si|Ve|Sr|Vr|Dr|Wr|Uw|Ge|Gp|1
+	 *                               |   \_/   \_/   \_/  |   \_/  |  |  |
+	 *		Di: Digest ------'    |     |     |   |    |   |  |  |
+	 *		En|De: Encrypt/Decrypt'     |     |   |    |   |  |  |
+	 *		Si|Ve: Sign/Verify ---------'     |   |    |   |  |  |
+	 *		Sr|Vr: SignUpdate/VerifyRecover --'   |    |   |  |  |
+	 *				Dr: Derive -----------'    |   |  |  |
+	 *				Wr|Uw: Wrap/Unwrap --------'   |  |  |
+	 *				Ge: Generate secret value -----'  |  |
+	 *				Gp: Generate secret pair ---------'  |
+	 *				1: One shot processing only ---------'
+	 */
+};
+
+#if CFG_TEE_TA_LOG_LEVEL > 0
+const char *mechanism_string_id(enum pkcs11_mechanism_id id)
+{
+	const size_t offset = sizeof("PKCS11_CKM_") - 1;
+	size_t n = 0;
+
+	for (n = 0; n < ARRAY_SIZE(pkcs11_modes); n++)
+		if (pkcs11_modes[n].id == id)
+			return pkcs11_modes[n].string + offset;
+
+	return "Unknown ID";
+}
+#endif /*CFG_TEE_TA_LOG_LEVEL*/
+
+/*
+ * Return true if @id is a valid mechanism ID
+ */
+bool mechanism_is_valid(enum pkcs11_mechanism_id id)
+{
+	size_t n = 0;
+
+	for (n = 0; n < ARRAY_SIZE(pkcs11_modes); n++)
+		if (id == pkcs11_modes[n].id)
+			return true;
+
+	return false;
+}
+
+/*
+ * Return true if mechanism ID is valid and flags matches PKCS#11 compliancy
+ */
+bool mechanism_flags_complies_pkcs11(uint32_t mechanism_type, uint32_t flags)
+{
+	uint32_t test_flags = 0;
+	size_t n = 0;
+
+	test_flags = flags & (PKCS11_CKFM_ENCRYPT | PKCS11_CKFM_DECRYPT |
```

Comment:
Why are these masked? Since the arguments are supposed to comply with PKCS#11 I don't see why unknown flags should be ignored. By the way, isn't the `_pkcs11` part rather implicit anyway?

Reply:
That's sanity to ensure we deal with flags that have a meaning.
This is currently used only upon debug, to validate what the TA will expose.

Reply:
Why should it be acceptable to supply flags without a meaning? Sounds like an error to me.

Reply:
This was to have a little of debug help here for one who may want to add a mechanism/algo. Using a table for PKCS#11 spec and a onther table for what is supported by the TA looked better to me. The function is caller from an `assert()`. Debug code checks that the added algo matches PKCS#11 spec.

Reply:
Do we keep this built-in safeguard of pkcs11 spec description?

Reply:
Sure, maybe only active while `assert()` is active.

Reply:
ok